src/obj_char_transfer.c

```diff
@@ -0,0 +1,872 @@
+
+void ObjCharTransfer_ReplaceGraphicsFromChardataByResID(int id, NNSG2dCharacterData *charData) {
+    GF_ASSERT(charData != NULL);
+    ObjCharTransferTask *sub = ObjCharTransfer_GetTaskByID(id);
+    GF_ASSERT(sub != NULL);
+    sub->charData = charData;
+    if (sub->vram & NNS_G2D_VRAM_TYPE_2DMAIN) {
+        GF_CreateNewVramTransferTask(NNS_GFD_DST_2D_OBJ_CHAR_MAIN, sub->baseAddrMain, charData->pRawData, charData->szByte);
+    }
+    if (sub->vram & NNS_G2D_VRAM_TYPE_2DSUB) {
+        GF_CreateNewVramTransferTask(NNS_GFD_DST_2D_OBJ_CHAR_SUB, sub->baseAddrSub, charData->pRawData, charData->szByte);
+    }
+}
+
+void ObjCharTransfer_ResetTransferTasksByResID(int resId) {
+    ObjCharTransferTask *task;
+    BOOL searchingCopies = TRUE;
+    do {
+        task = ObjCharTransfer_GetTaskByID(resId);
+        GF_ASSERT(task != NULL);
+        if (task->state == OBJ_CHAR_TRANSFER_TASK_COPY) {
+            ObjCharTransfer_DeleteTaskCopyByProxyPtr(&task->imageProxy);
+        } else {
+            searchingCopies = FALSE;
+        }
+    } while (searchingCopies);
+    if (task->state != OBJ_CHAR_TRANSFER_TASK_INIT) {
+        ObjCharTransferTask_Reset(task);
+        --sObjCharTransferTasksManager->num;
+    }
+}
+
+static void resetAllTransferTasks(void) {
+    for (int i = 0; i < sObjCharTransferTasksManager->max; ++i) {
+        if (sObjCharTransferTasksManager->tasks[i].state != OBJ_CHAR_TRANSFER_TASK_INIT) {
+            ObjCharTransferTask_Reset(&sObjCharTransferTasksManager->tasks[i]);
+            --sObjCharTransferTasksManager->num;
+        }
+    }
+}
+
+NNSG2dImageProxy *ObjCharTransfer_GetProxyPtrByResID(int resId) {
+    ObjCharTransferTask *sub = ObjCharTransfer_GetTaskByID(resId);
+    GF_ASSERT(sub != NULL);
+    if (sub->state == OBJ_CHAR_TRANSFER_TASK_INIT) {
+        return NULL;
+    }
+
+    return &sub->imageProxy;
+}
+
+NNSG2dImageProxy *ObjCharTransfer_ResizeTaskByResID_GetProxyPtr(int resId, u32 size) {
+    ObjCharTransferTask *task = ObjCharTransfer_GetTaskByID(resId);
+    GF_ASSERT(task != NULL);
+    if (task->state == OBJ_CHAR_TRANSFER_TASK_INIT) {
+        return NULL;
+    }
+    u32 offsetMain;
+    u32 offsetSub;
+    u32 newSizeMain;
+    u32 newSizeSub;
+    getBlockNumAndFreeSpaceForTransfer(task->vram, &offsetMain, &offsetSub, size, &newSizeMain, &newSizeSub);
+    if (task->state == OBJ_CHAR_TRANSFER_TASK_COPYABLE) {
+        return NULL;
+    }
+    task->state = OBJ_CHAR_TRANSFER_TASK_COPYABLE;
+    ObjCharTransferTask_UpdateBaseAddrs(task, offsetMain, offsetSub);
+    task->needResetBlockTransferFlags = TRUE;
+    task->sizeMain = newSizeMain;
+    task->sizeSub = newSizeSub;
+    ObjCharTransferTask_G2dLoadImageMappingVramTransfer(task);
+    reserveTransferBlocksByVramOffsetAndSize(task->vram, offsetMain, offsetSub, newSizeMain, newSizeSub);
+
+    return &task->imageProxy;
+}
+
+NNSG2dImageProxy *ObjCharTransfer_CopyTransferTaskByProxyPtr(NNSG2dImageProxy *proxyIn) {
+    ObjCharTransferTask *task;
```

Comment:
really dislike this c89 style local declaration, especially when they're then assigned in order as below, if you're gonna do this kinda thing at the top of the func pls at least declare and assign in the same part
it's 2024 ffs, this game was made in 2009, there's zero excuse to hold ourself to a standard made 20 years before the game was released, and 35 before now, especially when it doesn't affect matching, and we compile using c99

Reply:
(note this only applies to instances like the first two below)

Reply:
where you see this kind of forward declaration, just assume it's because declaration order/scope is necessary to match

Reply:
that is fair, however, the first assignment directly below to the first variable can (and should) be merged into one statement, and (probably) still match